tools/update.php

```diff
@@ -0,0 +1,746 @@
+<?php
+
+/**
+ * After running this script, a user should be able to go from an outdated LORIS
+ * instance to the newest install without issue.
+ *
+ * This includes having the latest minor release files, third-party packages
+ * (apt, npm, composer), PHP Version, etc.
+ *
+ * ## Details
+ *
+ * Specifically:
+ *
+ *- [x] **Creates a back up of `$LORIS_root`**
+ *
+ *- [x] (If not on a development instance) **Downloads the release source code
+ *      and overwrites `$LORIS_root`**
+ *
+ *- [x] **Ensures that apt dependencies are installed and up-to-date**
+ *
+ *- [x] **Runs 3rd-party package managers** (npm and composer)
+ *
+ *- [x] **Outputs a list of SQL patches** for the user to apply manually
+ *      (including development patches if on a LORIS dev instance)
+ *
+ *- [x] **Provides a lot of output** to assist the user in what's going on.
+ *
+ * Assumptions:
+ *
+ * As of now there is no "rollback" feature. `rsync` is used to overwrite the
+ * LORIS-root source code from the unzipped release folder. A backup is created,
+ * but should something go wrong with `rsync`, the user will need to restore the
+ * source code files manually. In this case, an error message is displayed and
+ * the script will exit.
+ *
+ * The function to retrieve patches is heavily dependent on the way we organize
+ * and name SQL files. If the directory structure changes, this functionality
+ * is likely to break.
+ *
+ * @category Update
+ * @package  Tools
+ * @author   John Saigle <john.saigle@mcgill.ca>
+ * @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ * @link     https://www.github.com/aces/Loris/
+ */
+
+require_once __DIR__ . "/../vendor/autoload.php";
+
+error_reporting(E_ALL);
+$required_major_php = 7; //TODO: Update these values as time passes
+$required_minor_php = 2;
+// PHP version required for LORIS.
+$php_version = "$required_major_php.$required_minor_php";
+if (PHP_MAJOR_VERSION < $required_major_php
+    || PHP_MINOR_VERSION < $required_minor_php
+) {
+    die("[-] ERROR: {$argv[0]} and LORIS require PHP v$php_version or higher.");
+}
+
+/* Create db connection and get version info. */
+$config    = \NDB_Config::singleton();
+$db_config = $config->getSetting('database');
+$db        = \Database::singleton(
```

Comment:
that might be problematic if the release patch requires admin credentials. (generally it shouldn't but all release patches have probably been run as admins to this day) just investigate

Reply:
I think it depends on the privileges of the SQL user, e.g. whether they can do ALTER, CREATE etc. That'll depend in a large part on the administrator of the system and how they set up their databases. 

I think in the worst case MySQL will give an error about not having the right permissions and the script will continue onward which shouldnt cause issues as patch application is the last step.

Reply:
it still places the DB is a half-update state and thats generally very difficult to recover front. A very Ideal way to tackle this would be to see what privileges are needed for a specific patch and verify if user have them.

the realistic way of doing this would be to add a notice before the run that says that all SQL patches are run as a regular user ORRR to use the quat credentials when/if they are available ?

Reply:
I implemented the "very Ideal way" :smile_cat:

Reply:
im very ideally impressed